### PR TITLE
New feature CommandEdit added with Junit testing

### PR DESCRIPTION
**Pull request by AjeyAshok**

### CODE REVIEW

File: src/main/java/executor/command/CommandEdit.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+import storage.StorageManager;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+
+public class CommandEdit extends Command {
+    protected String userInput;
```
Comment (Mudaafi) [inline src/main/java/executor/command/CommandEdit.java]: You don't have to declare this. It's declared in the abstract class Command already.

File: src/main/java/executor/command/CommandEdit.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        newValue = Parser.parseForFlag("value", this.userInput);
+        newDate = Parser.parseForFlag("date", this.userInput);
+        if (newTag != null) {
+            updateTag(storageManager);
+        } else if (newValue != null) {
+            updateCashValue(storageManager);
+        } else if (newDate != null) {
+            updateDate(storageManager);
+        } else {
+            throw new DukeException("Flag invalid. Valid input : tag/value/date");
```
Comment (Mudaafi) [inline src/main/java/executor/command/CommandEdit.java]: Shouldn't it be /tag <tag> /value <value> /date <dd/MM/yy> ?

File: src/main/java/executor/command/CommandEdit.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        return "\nNOTE : Tag contains other characters apart from alphabets";
+    }
+
+    /**
+     * Function to output a String message if the tag input by user is empty.
+     * @return is the output message
+     */
+    private String noteTagEmpty() {
+        return "\nNOTE : Tag is empty";
+    }
```
Comment (Mudaafi) [inline src/main/java/executor/command/CommandEdit.java]: abstraction is good and all but couldn't you have just set a variable to equal to the string HAHA

File: src/main/java/executor/command/CommandEdit.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        changeTag.add(newTag);
+        oldTag = storageManager.getWallet().getReceipts().get(index).getTags();
+        storageManager.getWallet().getReceipts().get(index).setTags(changeTag);
+        if (newTag.isEmpty()) {
+            outputMessageOnGui(messageForTagChange() + noteTagEmpty());
+        } else if (isAlpha(newTag)) {
+            outputMessageOnGui(messageForTagChange());
+        } else {
+            outputMessageOnGui(messageForTagChange() + noteHasNonAlphabets());
+        }
```
Comment (Mudaafi) [inline src/main/java/executor/command/CommandEdit.java]: Doesn't this reset the infoCapsule's output string?

File: src/main/java/executor/command/CommandEdit.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            }
+            changeValue = Double.parseDouble(newValue);
+            checkAndChangeToTwoDP();
+            oldValue = storageManager.getWallet().getReceipts().get(index).getCashSpent();
+            storageManager.getWallet().getReceipts().get(index).setCashSpent(roundedChangeValue);
+            if (moreThanTwoDP) {
+                outputMessageOnGui(messageForValueChange() + noteMoreThanTwoDP());
+            } else {
+                outputMessageOnGui(messageForValueChange());
+            }
```
Comment (Mudaafi) [inline src/main/java/executor/command/CommandEdit.java]: Doesn't this reset the infoCapsule's output string?

File: src/main/java/executor/command/CommandEdit.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    private void updateDate(StorageManager storageManager) {
+        try {
+            changeDate = LocalDate.parse(newDate);
+            checkIfDateIsInFuture();
+            oldDate = storageManager.getWallet().getReceipts().get(index).getDate();
+            storageManager.getWallet().getReceipts().get(index).setDate(changeDate);
+            if (isFutureDate) {
+                outputMessageOnGui(messageForDateChange() + noteFutureYear());
+            } else {
+                outputMessageOnGui(messageForDateChange());
```
Comment (Mudaafi) [inline src/main/java/executor/command/CommandEdit.java]: Doesn't this reset the infoCapsule's output string?

File: src/main/java/executor/command/CommandEdit.java (hunk 1 above)
Comment (AjeyAshok) [reply src/main/java/executor/command/CommandEdit.java]: Noted

**Pull request merged**